pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -829,9 +864,112 @@ func (n *NetworkPolicyController) deletePod(old interface{}) {
+	for group := range appliedToGroupKeys {
+		n.enqueueAppliedToGroup(group)
+	}
+	for group := range addressGroupKeys {
+		n.enqueueAddressGroup(group)
+	}
+}
+
+// addExternalEntity retrieves all AddressGroups and AppliedToGroups which match the ExternalEnitty's
+// labels and enqueues the groups key for further processing.
+func (n *NetworkPolicyController) addExternalEntity(obj interface{}) {
+	defer n.heartbeat("addExternalEntity")
+	ee := obj.(*v1alpha1.ExternalEntity)
+	klog.V(2).Infof("Processing External Entity %s/%s ADD event, labels: %v", ee.Namespace, ee.Name, ee.Labels)
+	// Find all AppliedToGroup keys which match the ExternalEntity's labels.
+	appliedToGroupKeySet := n.filterAppliedToGroupsForPodOrExternalEntity(ee)
+	// Find all AddressGroup keys which match the ExternalEntity's labels.
+	addressGroupKeySet := n.filterAddressGroupsForPodOrExternalEntity(ee)
+	// Enqueue groups to their respective queues for group processing.
+	for group := range appliedToGroupKeySet {
+		n.enqueueAppliedToGroup(group)
+	}
+	for group := range addressGroupKeySet {
+		n.enqueueAddressGroup(group)
+	}
+}
+
+// updateExternalEntity retrieves all AddressGroups and AppliedToGroups which match the
+// updated and old ExternalEntity's labels and enqueues the group keys for further
+// processing.
+func (n *NetworkPolicyController) updateExternalEntity(oldObj, curObj interface{}) {
+	defer n.heartbeat("updateExternalEntity")
+	oldEE := oldObj.(*v1alpha1.ExternalEntity)
+	curEE := curObj.(*v1alpha1.ExternalEntity)
+	klog.V(2).Infof("Processing ExternalEntity %s/%s UPDATE event, labels: %v", curEE.Namespace, curEE.Name, curEE.Labels)
+	// No need to trigger processing of groups if there is no change in the
+	// Pod labels or Pods Node or Pods IP.
+	labelsEqual := labels.Equals(labels.Set(oldEE.Labels), labels.Set(curEE.Labels))
+	// TODO: Right now two ExternalEntities are only considered equal if the list of Endpoints and
+	//  all NamedPorts in each Endpoint are of the exact order. This constraint might be too strict.
+	if labelsEqual && cmp.Equal(oldEE.Spec, curEE.Spec) {
+		klog.V(4).Infof("No change in ExternalEntity %s/%s. Skipping NetworkPolicy evaluation.", curEE.Namespace, curEE.Name)
+		return
+	}
+	// Find groups matching the old ExternalEntity's labels.
+	oldAddressGroupKeySet := n.filterAddressGroupsForPodOrExternalEntity(oldEE)
+	oldAppliedToGroupKeySet := n.filterAddressGroupsForPodOrExternalEntity(oldEE)
+	// Find groups matching the new ExternalEntity's labels.
+	curAppliedToGroupKeySet := n.filterAddressGroupsForPodOrExternalEntity(curEE)
+	curAddressGroupKeySet := n.filterAddressGroupsForPodOrExternalEntity(curEE)
+	// Create set to hold the group keys to enqueue.
+	var appliedToGroupKeys sets.String
+	var addressGroupKeys sets.String
+	// AppliedToGroup keys must be enqueued only if the ExternalEntity's spec has changed or
+	// if ExternalEntity's label change causes it to match new Groups.
+	if !cmp.Equal(oldEE.Spec, curEE.Spec) {
```

Comment:
wondering  if reflect.DeepEqual does not work here?

Reply:
Done and added TODO.